### PR TITLE
Created Three Optimization Objectives

### LINKED ISSUE #24: Create multiple tests for the different Distance optimizers
1. How do we test the distance optimizers?
    a. See how each of the methods work
    b. We are looking for the differences of the distance between two points for all
the selected methods. The points distance will range between 1m, 10m, 100m,
1km, 5km, 10km, 50km and 100km.
2. Methods that we are going to test
    a. Test the OMPL's distance objective (Raye's method)
    b. Test the Euclidean Distance method
    c. Test the Thunderbird's distance method
    d. Check and test whether Manhattan or Minkowski Distance
    
All the test process, data, results and evaluation will be rendered in confluence and will be separated in files as reference.
    
Further Information is in ["How Do We Calculate the Optimization Objectives?"](https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1860403201/How+Do+We+Calculate+the+Optimization+Objectives)
https://ompl.kavrakilab.org/optimizationObjectivesTutorial.html

### LINKED ISSUE #25: Create a Heading Objective
1. What is defined as "better"? A better objective method is one where it improves
one or all of the following criteria:
    a. It can produce smaller tacks
    b. There are no issues that can lead to the Sailbot to get damaged when there is an obvious safer path.
2. We have two methods to test:
    a. Is it better to measure the angle from the previous position (Raye's method)
    b. Is it better to measure the angle from the final position (A heuristic, my method)
 
All the test process, data, results and evaluation will be rendered in confluence and will be separated in files as reference.

Further Information is in ["How Do We Calculate the Optimization Objectives?"](https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1860403201/How+Do+We+Calculate+the+Optimization+Objectives)

### LINKED ISSUE #46: Tests for coord_systems.py
### Purpose

Make sure all functions in [coord_systems.py](https://github.com/UBCSailbot/local_pathfinding/blob/main/local_pathfinding/coord_systems.py) are tested

### Resources

- tests file: https://github.com/UBCSailbot/local_pathfinding/blob/main/test/test_coord_systems.py

### PR DESCRIPTION
**Pull request by FireBoyAJ24**
### Description

- #25 
- #24 

### Resolves  #25

I have added the distance, minimum turning and upwind/downwind optimization objectives. We can use this as a starting point to build upon, test and verify our optimization objectives.

Note: Two test cases were made for each of the test functions. Later we can add more test cases in a larger issue.

### Verification

- [x] Check whether all the distance objective functions are not producing an error.
- [x] Check whether all the minimum turning objective functions are not producing an error.
- [x] Check whether the wind objective function is not producing an error
- [x] Ensure there is no unintended code in the local path and ompl path files

### Resources

- https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1860403201/How+Do+We+Calculate+the+Optimization+Objectives

Note: Please ignore #41 and #32 as this pull request replaces it.

### CODE REVIEW

**Comment by patrick-5546** [comment]
> Note: Test cases have not been made as requested in https://github.com/UBCSailbot/local_pathfinding/issues/24 since that is a separate major issue to tackle.

If this is the case does this PR resolve the issue, or should we leave it open even after merging in this PR?

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Don't we want to raise a `NontImplementedError` if an objective doesn't implement `motionCost`?

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py:46]: Did not copy over type hints

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Move attribute documentation to `Objective`, where it is set

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Redundant, as it is set in `Objective`

File: test/test_objectives.py
Comment (patrick-5546) [inline test/test_objectives.py]: Parameterize

File: test/test_objectives.py
Comment (patrick-5546) [inline test/test_objectives.py]: Parameterize

**Review by patrick-5546 — COMMENTED** [review]
General comments: look at the files changed tab to see the latest changes from the skeleton code PR that are in main that were not applied to this branch
- Add type hints to all functions except for `simple_setup` and `space_information` arguments
- Function naming convention: only OMPL functions should be camelCase, everything else should be snake_case (no need to start with `_`)

**Comment by FireBoyAJ24** [comment]
> > Note: Test cases have not been made as requested in #24 since that is a separate major issue to tackle.
> 
> If this is the case does this PR resolve the issue, or should we leave it open even after merging in this PR?

The PR resolves the basic  testing of each function. I want to have a larger issue that requests the addition of more test cases.

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Could add an attribute `implementation` to select between the different implementations

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Different units, bounding radians to 180 degrees

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Document attribute `wind_direction` We don't need to include units in variable name, but it should be stated in docstrings.

Comment (patrick-5546) [reply local_pathfinding/objectives.py]: Also I think if you make this radians you don't have to convert anywhere else in `WindObjective`

**Review by patrick-5546 — COMMENTED** [review]
I fixed the rest of the inconsistencies with the main branch. In the future, you can avoid this by double checking the files changed tab.

One other note is that we need to be able to update the objective attributes. For example, `wind_direction` or `goal_x`, `goal_y` will change during runtime.

And just a reminder to pull this branch to get my changes before making new changes to avoid merge conflicts.

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Document `implementation` attribute for both distance and minimum turning objectives

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: else raise `ValueError(f"Implementation {self.implementation} not supported")` for both distance and minimum turning objectives

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Best way would be to create an enum type for DistanceMethod and TurnCost

For example
```py
from enum import Enum, auto

class Color(Enum):
    RED = auto()
    BLUE = auto()
    GREEN = auto()
```

Resource: https://docs.python.org/3/howto/enum.html

Comment (patrick-5546) [reply local_pathfinding/objectives.py]: Enums can go at the top of this file

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Range is [0, 2pi)

https://github.com/UBCSailbot/custom_interfaces/blob/main/msg/HelperHeading.msg

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Range is [-pi, pi)

https://github.com/UBCSailbot/custom_interfaces/blob/main/msg/WindSensor.msg

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Have you thought about the bounds of each variable? You may have to account for crossing the 2pi/0 line. And wouldn't you want to get the smaller of the angles between them?

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Can change to `UPWIND_MAX_ANGLE_RADIANS = math.radians(40.0)`

File: local_pathfinding/objectives.py
Comment (patrick-5546) [inline local_pathfinding/objectives.py]: Any reason why they override the default `implementation`s?

**Review by patrick-5546 — COMMENTED** [review]
Next round of comments

File: local_pathfinding/objectives.py
Comment (FireBoyAJ24) [reply local_pathfinding/objectives.py]: Just because.....

File: local_pathfinding/objectives.py
Comment (FireBoyAJ24) [reply local_pathfinding/objectives.py]: When I use `math.atan2(x, y)` then my angles gets bounded between (-pi, pi]. (Note: I am saying `math.atan2(x, y)` and not `math.atan2(y, x)`)

File: local_pathfinding/coord_systems.py
Comment (patrick-5546) [inline local_pathfinding/coord_systems.py:77]: @jamenkaye will be writing tests for this

Comment (jamenkaye) [reply local_pathfinding/coord_systems.py:77]: Added tests!

**Comment by jamenkaye** [comment]
Good thing I added tests for `xy_to_latlon()`, I caught a bug where `true_bearing` was converted to degrees twice :)

@FireBoyAJ24 I noticed `test_get_latlon_path_length_objective()` was failing too. Any idea what's up with this? I looked for a little while but couldn't find the issue.

**Comment by jamenkaye** [comment]
Adding in this information from Jay's entry in the Pathfinding meeting notes:

I was integrating the `xy_to_lat-lon` function (coord_systems.py) into the `_get_latlon_path_length_objective` (objectives.py) function.

Issue: My expected distance between two lat-lon coordinates does not agree with the calculated distance from GEODESIC. In short `test_get_latlon_path_length_objectivefunction` (test_objectives.py) is spitting out a not expected error.

Other than that we need to do another round of review to ensure all the formatting is corrected and all the tests have passed.

File: local_pathfinding/objectives.py
Comment (jamenkaye) [inline local_pathfinding/objectives.py]: Really these are minimum angles, right? 
We want to communicate that we can't sail closer to irons than 40 degrees. Maybe `HIGHEST_UPWIND_ANGLE` and `LOWEST_DOWNWIND_ANGLE` would be more descriptive

File: local_pathfinding/objectives.py
Comment (jamenkaye) [inline local_pathfinding/objectives.py]: more specifically, the comment could say `# Calculate the true bearing of s2 from s1`

File: local_pathfinding/objectives.py
Comment (jamenkaye) [inline local_pathfinding/objectives.py]: similarly, this could say `# Calculate the true bearing of the goal from s1`

File: local_pathfinding/objectives.py
Comment (jamenkaye) [inline local_pathfinding/objectives.py:265]: I could be wrong but it seems like the `fabs` function isn't needed in this line? In the `if` block, `turn_size_unbias` is $2 \pi - \text{ angle between } (\pi, 2\pi]$, and in the else block, `turn_size_unbias` is already positive due to the `fabs` on line 246.

File: local_pathfinding/objectives.py
Comment (jamenkaye) [inline local_pathfinding/objectives.py]: how about `# Calculate the true bearing of the goal from s1` ?

File: local_pathfinding/objectives.py
Comment (jamenkaye) [inline local_pathfinding/objectives.py]: same thing, but how about `# Calculate the true bearing of s2 from s1`?

File: local_pathfinding/objectives.py
Comment (jamenkaye) [inline local_pathfinding/objectives.py]: I saw you were using a hypotenuse function elsewhere, should that maybe be used here?

File: local_pathfinding/objectives.py
Comment (jamenkaye) [inline local_pathfinding/objectives.py:354]: for Jamen: Does this work for unbounded angles? the `is_upwind` function could give unbounded angles.

**Review by jamenkaye — COMMENTED** [review]
Reviewed!

File: local_pathfinding/objectives.py
Comment (jamenkaye) [reply local_pathfinding/objectives.py:354]: Made `is_angle_between()` handle un bounded angles, because of functions like `is_upwind()`. In `is_upwind()`, we could have `wind_angle` = pi (or nearly pi), then when we add on the `HIGHEST_UPWIND_ANGLE_RADIANS` we could pass an unbounded angle to `is_angle_between()`.

File: local_pathfinding/objectives.py
Comment (jamenkaye) [reply local_pathfinding/objectives.py]: This works though, its fine 👍

File: local_pathfinding/objectives.py
Comment (jamenkaye) [reply local_pathfinding/objectives.py:265]: Since an extra fabs() function is such a small inefficiency, but taking it out could possibly risk breaking something, I'll just leave it in.

**Review by jamenkaye — APPROVED** [review]
Looks good!

I think there might be an unecessary `fabs()` function on line 256: `return math.degrees(math.fabs(turn_size_unbias))`, but this is a very small inefficiency and I don't want to risk breaking anything haha

**Comment by jamenkaye** [comment]
... But I'd still love a second opinion :) @patrick-5546

**Comment by patrick-5546** [comment]
Summarizing the changes I made
- Fix types and bounds
- Refactor to reduce functions/tests that depend on OMPL

**Pull request merged**
